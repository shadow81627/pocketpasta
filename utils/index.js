/**
 * Remove all specified keys from an object, no matter how deep they are.
 * The removal is done in place, so run it on a copy if you don't want to modify the original object.
 * This function has no limit so circular objects will probably crash the browser
 *
 * @param obj The object from where you want to remove the keys
 * @param keys An array of property names (strings) to remove
 */
function removeKeys(obj, keys) {
  for (const property in obj) {
    // important check that this is objects own property
    // not from prototype prop inherited
    const hasProperty = Object.prototype.hasOwnProperty.call(obj, property);
    if (hasProperty) {
      switch (typeof obj[property]) {
        case 'string':
          if (keys.includes(property)) {
            delete obj[property];
          }
          break;
        case 'object':
          if (keys.includes(property)) {
            delete obj[property];
          } else {
            removeKeys(obj[property], keys);
          }
          break;
      }
    }
  }
}

export { removeKeys };
