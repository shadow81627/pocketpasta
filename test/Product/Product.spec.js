import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import BootstrapVuePlugin from 'bootstrap-vue';
import { removeKeys } from '@/utils';
import Component from '@/components/Product/Product';

import product from '@/assets/link-data/products/tomatoes.json';

// JSDOM complains about this attribute
removeKeys(product, ['@type']);

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(BootstrapVuePlugin);

const $route = {
  path: '/products/detail/1',
  params: { id: '1' },
  fullPath: '/products/detail/1',
  name: 'products-detail-id',
};

describe('Product', () => {
  let getters;
  let state;
  let store;
  let mutations;

  beforeEach(() => {
    state = {
      products: [{ ...product, id: 1 }, { id: 2 }, { id: 3 }, { id: 4 }],
    };

    getters = {
      getProductById: (state) => (id) =>
        state.products.find((product) => product.id === id),
    };

    mutations = {
      setTheme(currentState, theme) {
        currentState.theme = theme;
      },
    };

    store = new Vuex.Store({
      state,
      getters,
      mutations,
    });
  });

  const factory = () =>
    shallowMount(Component, {
      store,
      localVue,
      mocks: {
        $route,
      },
    });

  test('mounts properly', () => {
    const wrapper = factory();
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test('renders properly', () => {
    const wrapper = factory();
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('head', () => {
    expect(Component.head()).toBeTruthy();
  });
});
