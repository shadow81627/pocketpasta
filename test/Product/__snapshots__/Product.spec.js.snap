// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe renders properly 1`] = `"<div class=\\"product\\"><h1>Audrey Hepburn’s Spaghetti Al Pomodoro Recipe</h1> <p>Make Audrey Hepburn's spaghetti al pomodoro recipe. Hepburn liked her pasta smothered in tomato sauce.</p> <img src=\\"https://img.buzzfeed.com/thumbnailer-prod-us-east-1/video-api/assets/162208.jpg\\" alt=\\"Audrey Hepburn’s Spaghetti Al Pomodoro Recipe\\" class=\\"img-fluid mx-auto d-block\\"> <share-stub class=\\"py-4\\"></share-stub> <p><strong>Servings:</strong> <span>4 servings</span></p> <h2>Ingredients:</h2> <ol class=\\"list-group-flush\\"></ol> <!----> <keywords-stub label=\\"Tags\\" tags=\\"Audrey Hepburn, recipe, Pasta Recipes, Sauce Recipes, Audrey, Hepburn, spaghetti, spathetti, pomodoro, pasta, history, sauce\\"></keywords-stub></div>"`;
